### PR TITLE
Add code

### PR DESCRIPTION
**Pull request by SweetOps**
## What
* Add code

## Why
* Need module for named `subnets`

### CODE REVIEW

**Comment by SweetOps** [comment]
screenshot 2017-10-19 16 19 51

File: main.tf [hunk 1]
```diff
@@ -0,0 +1,33 @@
+module "vpc" {
```
Comment (osterman) [inline main.tf]: This is a subnet module. Does not make sense to ship a `vpc` with it.

**Comment by osterman** [comment]
- [ ] drop the concept of public/private
- [ ] add a flag to enable nat gateways
- [ ] make the subnets named (e.g. user inputs a list of names, and it creates a list of subnets with those names)

**Comment by osterman** [comment]
Interface should be
```
module "subnets" {
  namespace = "${var.namespace}"
  stage = "${var.stage}"
  names = ["charlie", "echo", "bravo"]
  vpc_id = "vpc-1234"
  base_cidr = "10.0.1.0./24"
  nat_enabled = "true" 
}
```

**Comment by SweetOps** [comment]
screenshot 2017-10-20 17 03 56

File: main.tf [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  id = "${var.vpc_id}"
+}
+
+module "subnets_label" {
+  source     = "git::https://github.com/cloudposse/terraform-null-label.git?ref=tags/0.2.2"
+  namespace  = "${var.namespace}"
+  name       = "${var.name}"
+  stage      = "${var.stage}"
+  delimiter  = "${var.delimiter}"
+  attributes = ["subnet"]
```
Comment (osterman) [inline main.tf]: Merge with `attributes` input

File: main.tf [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  tags = {
+    "Name"      = "${module.subnets_label.id}${var.delimiter}${element(var.names, count.index)}${var.delimiter}${element(var.availability_zones, count.index)}"
+    "Stage"     = "${module.subnets_label.stage}"
+    "Namespace" = "${module.subnets_label.namespace}"
+  }
+}
+
+## NGW
+resource "aws_eip" "default" {
+  count = "${var.nat_enabled ? length(compact(var.names)) * length(compact(var.availability_zones)) : 0}"
```
Comment (osterman) [inline main.tf]: Use a `local` variable for `length(compact(var.names)) * length(compact(var.availability_zones)`

File: main.tf [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+resource "aws_route_table" "default" {
+  count  = "${length(compact(var.names)) * length(compact(var.availability_zones))}"
+  vpc_id = "${var.vpc_id}"
+  tags   = "${module.subnets_label.tags}"
+}
+
+resource "aws_route" "default" {
+  count                  = "${length(compact(var.names)) * length(compact(var.availability_zones))}"
+  route_table_id         = "${element(aws_route_table.default.*.id, count.index)}"
+  nat_gateway_id         = "${replace(element(coalescelist(aws_nat_gateway.default.*.id, list("workaround")), count.index), "workaround", "")}"
```
Comment (osterman) [inline main.tf]: What is the `workaround` for?

File: README.md [hunk 5]
```diff
@@ -1 +1,61 @@
...
+| `base_cidr`          |   ``    | The base CIDR block which will be divided into subnet CIDR blocks (e.g. `10.0.0.0/16`)                              |    No    |
+| `igw_id`             |   ``    | The Internet Gateway ID public route table will point to (e.g. `igw-9c26a123`). Conflicts with `nat_enabled = true` |    No    |
+| `nat_enabled`        | `false` | Flag to enable/disable NAT gateways. Conflicts with non-empty `igw_id`                                              |    No    |
+| `availability_zones` |   ``    | The  list of Availability Zones where subnets will be created (e.g. `["us-east-1a", "us-east-1b"]`).                |   Yes    |
+| `network_acl_id`     |   ``    | Network ACL ID that will be added to the subnets.  If empty, a new ACL will be created                              |    No    |
+| `additional_routes`  |  `{}`   | Map of Additional routes for subnets (e.g. `{"10.0.0.2/24" = igw-0038f979}`)                                        |    No    |
+| `egress`             |  `[]`   | Specifies an egress rules                                                                                           |    No    |
```
Comment (osterman) [inline README.md]: `egress_acl` .... `Specifies a list of CIDRs permitted egress`

File: README.md [hunk 6]
```diff
@@ -1 +1,61 @@
...
+| `ingress`            |  `[]`   | Specifies an ingress rule                                                                                           |    No    |
```
Comment (osterman) [inline README.md]: `ingress_acl` .... `List of CIDRs permitted ingress`

File: outputs.tf [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  value       = ["${aws_nat_gateway.default.*.public_ip}"]
+  description = "The public IP addresses of the NAT Gateways"
+}
+
+output "subnet_ids" {
+  value       = ["${aws_subnet.default.*.id}"]
+  description = "IDs of subnets"
+}
+
+module "private_subnets" {
```
Comment (osterman) [inline outputs.tf]: Copy & paste error? Do not include module in outputs

File: outputs.tf [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  namespace          = "${var.namespace}"
+  stage              = "${var.stage}"
+  names              = ["charlie", "echo", "bravo"]
+  vpc_id             = "vpc-1234"
+  base_cidr          = "10.0.1.0.1/24"
+  availability_zones = ["us-east-1a", "us-east-1b"]
+  nat_enabled        = "true"
+}
+
+module "public_subnets" {
```
Comment (osterman) [inline outputs.tf]: Copy paste error?

File: outputs.tf (hunk 7 above)
Comment (SweetOps) [reply outputs.tf]: IDE issue :)

File: main.tf (hunk 4 above)
Comment (SweetOps) [reply main.tf]: terraform doesn't allow for executing element(list, index) on empty lists

File: main.tf [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  tags = {
+    "Name"      = "${module.subnets_label.id}${var.delimiter}${element(var.names, count.index)}${var.delimiter}${element(var.availability_zones, count.index)}"
+    "Stage"     = "${module.subnets_label.stage}"
+    "Namespace" = "${module.subnets_label.namespace}"
+  }
+}
+
+resource "aws_eip" "default" {
+  count = "${var.nat_enabled ? local.subnets_count : 0}"
```
Comment (osterman) [inline main.tf]: we should explicitly evaluate against `== "true"`

Comment (osterman) [reply main.tf]: ....everywhere we have a ternary

**Comment by osterman** [comment]
- [ ] add CI See this for example: https://github.com/cloudposse/terraform-null-label/pull/9/files

File: README.md [hunk 10]
```diff
@@ -1 +1,61 @@
-# terraform-aws-named-subnets
\ No newline at end of file
+# terraform-aws-named-subnets
+
+Terraform module for named `subnets` provisioning.
```
Comment (osterman) [inline README.md]: Perhaps we should link to this for an explanation of the math: https://github.com/cloudposse/terraform-aws-dynamic-subnets#tldr

File: variables.tf [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+  type    = "map"
+  default = {}
+}
+
+variable "network_acl_id" {
+  description = "Network ACL ID that will be added to the subnets. If empty, a new ACL will be created "
+  default     = ""
+}
+
+variable "egress" {
```
Comment (osterman) [inline variables.tf]: Rename to `network_acl_egress`

File: variables.tf [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+      action     = "allow"
+      cidr_block = "0.0.0.0/0"
+      from_port  = 0
+      to_port    = 0
+      protocol   = "-1"
+    },
+  ]
+}
+
+variable "ingress" {
```
Comment (osterman) [inline variables.tf]: Rename to `network_acl_ingress`

File: README.md [hunk 13]
```diff
@@ -29,23 +29,23 @@ module "public_subnets" {
...
-| `egress`             |  `[]`   | List of Network ACL CIDRs permitted egress                                                                          |    No    |
-| `ingress`            |  `[]`   | List of Network ACL CIDRs permitted ingress                                                                         |    No    |
+| Name                  | Default | Description                                                                                                         | Required |
+|:----------------------|:-------:|:--------------------------------------------------------------------------------------------------------------------|:--------:|
+| `namespace`           |   ``    | Namespace (e.g. `cp` or `cloudposse`)                                                                               |   Yes    |
+| `stage`               |   ``    | Stage (e.g. `prod`, `dev`, `staging`)                                                                               |   Yes    |
+| `name`                |   ``    | Name  (e.g. `bastion` or `db`)                                                                                      |   Yes    |
```
Comment (osterman) [inline README.md]: How is name used now? Seems like we should drop it.

File: main.tf [hunk 14]
```diff
@@ -25,7 +25,7 @@ resource "aws_subnet" "default" {
   cidr_block        = "${cidrsubnet(length(var.base_cidr) > 0 ? var.base_cidr : data.aws_vpc.default.cidr_block, ceil(log(local.subnets_count * 2, 2)), count.index)}"
 
   tags = {
-    "Name"      = "${module.subnets_label.id}${var.delimiter}${element(var.names, count.index)}${var.delimiter}${element(var.availability_zones, count.index)}"
+    "Name"      = "${module.subnets_label.namespace}-${module.subnets_label.stage}-${format("%v-%v", element(var.names, count.index, element(var.availability_zones, count.index))}"
```
Comment (osterman) [inline main.tf]: What about the other `tags`? anywhere we use `subnets_count` should be the same

File: README.md (hunk 13 above)
Comment (SweetOps) [reply README.md]: `name` is required field for `terraform-null-label`

File: main.tf (hunk 14 above)
Comment (SweetOps) [reply main.tf]: I've changed back `module.subnets_label.namespace` to `module.subnets_label.id`, lets do all cosmetic fixes in next release.

File: README.md [hunk 15]
```diff
@@ -1 +1,65 @@
...
+module "private_subnets" {
+  source             = "git::https://github.com/cloudposse/terraform-aws-named-subnets.git?ref=master"
+  name               = "${var.name}"
+  namespace          = "${var.namespace}"
+  stage              = "${var.stage}"
+  names              = ["charlie", "echo", "bravo"]
+  vpc_id             = "vpc-1234"
+  base_cidr          = "10.0.1.0/24"
+  availability_zones = ["us-east-1a", "us-east-1b"]
+  nat_enabled        = "true"
```
Comment (comeanother) [inline README.md]: I won't work properly for `private` subnets, because NAT GW should be in `public` subnet.

Comment (SweetOps) [reply README.md]: NAT GW can't be created without EIP
https://www.terraform.io/docs/providers/aws/r/nat_gateway.html

Comment (SweetOps) [reply README.md]: or, what you mean?

Comment (osterman) [reply README.md]: @comeanother  is referring to the fact that nat gateways only work in public subnets, so the caller must be aware that they can only use the `nat_gateway` flag on the private subnet. it only works if there's an internnet gateway associated with the subnet.

File: README.md (hunk 13 above)
Comment (osterman) [reply README.md]: if you set name to `""` it will drop out.

Comment (osterman) [reply README.md]: or actually, i would use `%v` and then it's a `format` template.

File: README.md (hunk 15 above)
Comment (SweetOps) [reply README.md]: @osterman, thats out of date, module has been reworked

**Comment by osterman** [comment]
Closed in favor of https://github.com/cloudposse/terraform-aws-named-subnets/pull/3

**Pull request closed without merging**